numpy/core/src/multiarray/ctors.c

```diff
@@ -784,6 +784,33 @@ PyArray_NewFromDescr_int(
 
 
     if (data == NULL) {
+        NPY_traverse_info fill_zero_info;
+        /* float errors do not matter and we do not release GIL */
+        NPY_ARRAYMETHOD_FLAGS zero_flags = PyArrayMethod_MINIMAL_FLAGS;
+        NPY_traverse_info_init(&fill_zero_info);
```

Comment:
We need to pair this with the free/clear function, unfortunately (this will be important for structured dtypes).

One path may be to not support/check for `fill_zero_info.func == NULL` here, but only use `get_fill_zero_loop == NULL` to make the decision about calloc.  That would move this whole logic down to after the calloc choice was made.  (structured dtypes won't be able to signal calloc is good enough, but :shrug:)

Alternatively, we just need to add the cleanup either in `fail:` of before all relevant goto fails.

Reply:
Oh oops, I missed the existence of `NPY_traverse_info_xfree` looking at the traversal code, thank you very much for catching this.

Reply:
I ended up refactoring so the cleanup goes in the `fail` block. This does lead to allocating and initializing the traverse info struct in the fast path, but it's a tiny struct so I don't think that's a problem. Let me know if you'd like further massaging on this.